Use `npm repo` to open javascript packages

jdsutherland opened the pull request:
Closes #4

jdsutherland committed:
Use `npm repo` to open javascript packages

gabebw commented on plugin/github_link_opener.vim:
```diff
@@ -12,9 +12,12 @@ function! s:OpenGitHubLink()
   if &ft ==# 'go' && has_more_than_one_slash
     let url = matchstr(word, chars . '/' . chars . '/' . chars)
     call s:OpenWithNetrw("https://" . url)
+  elseif &ft ==# 'javascript' && line_has_js_package || &ft ==# 'json'
```
For the right side of this line (`&ft ==# 'json'`), I'm concerned that it will lead to unexpected behavior, since this will override the default behavior of <kbd>gx</kbd> for all JSON files. Imagine a JSON file that's not `package.json` that has a URL in it. This line means that <kbd>gx</kbd> will try to run `npm repo` on that `<cword>`, rather than just opening it as usual.

For example, on this URL from a real `package.json`, I ran <kbd>gx</kbd> with my cursor on the `https`, with the changes from this PR:

```json
"bugs": {
  "url": "https://github.com/form-data/form-data/issues"
},
```

Unfortunately, it didn't work correctly because it only grabbed the `cword`, and so since the cursor was on `https`, it ran `npm repo https` and showed an error.

Overall, I think that:

1. Running `npm repo` when `line_has_js_package` is true is quite safe, and a welcome change. Thank you! 🙇 
2. Running it in a json file, even a `package.json` file, is tricky and will likely lead to unwanted behavior for end users. ⚠️ 

Given that, I think that I'd prefer changing this line to something like this and just not do anything in JSON files:

```vimscript
elseif index(['javascript', 'javascript.jsx', 'typescript'], &ft) >= 0 && line_has_js_package
```

I believe that `import`/`require` statements could also appear in Typescript (ft == `typescript`) and JSX (ft == `javascript.jsx`), not just `javascript`, so I added a few more cases.

(Note that we can now get away with not thinking about `==#` vs `==` because `index()` is case-sensitive always, even if `ignorecase` is on. Phew!)

What do you think?

jdsutherland replied:
I agree about the json issues - this was a half-baked idea. I'd expect most usage would come from reading js-like files anyway so dropping it doesn't seem like much of a loss.

I also agree that covering Typescript and React makes sense. I'm not fully steeped in modern javascript land yet so thanks for pointing these out.

I'll make the changes.

jdsutherland replied:
I just realized that imports like
`import { filter } from 'rxjs/operators';` where `<cword>` from cursor on `operators` would pass `operators` rather than `rxjs` leading to unwanted behavior. I think we want to only use the first word before a slash. I'll try using a regex capture to get this working.

gabebw reviewed (commented):
Thanks so much for making a PR (and bonus points for updating the README – it's much appreciated!).

I left a comment about expanding the scope to other Javascript-ish files, plus some funky stuff that could happen in JSON files - let me know what you think.

jdsutherland committed:
Remove json, add React, Typescript fts

jdsutherland committed:
Add 'from' for matching js packages

from ensures multiline imports are usable:
import {
  forceCollide,
  forceLink,
  zoom,
} from 'd3';

jdsutherland committed:
Extract package from line w/ regex

Using getline is better UX and validates input passed to `npm
repo`, avoiding local files.

regex matches:
"some-package"
"example.com"
"under_score"
"123numeric"
"@npm/thingy"
"@jane/foo.js"

jdsutherland commented:
I ran through a couple of popular codebases and  found a few more issues but I think things are now in order. I made it so that the cursor position doesn't matter as it just parses the name from `getline` using https://github.com/npm/validate-npm-package-name#valid-names as a source for the regex - see https://regex101.com/r/r3Q6zS/1. 

I learned that npm has something called [scopes](https://docs.npmjs.com/misc/scope) and it seems that grabbing the first name before the slash won't always link to the correct repo (e.g. @angular/cli would link to https://github.com/angular/angular but actual is https://github.com/angular/angular-cli) but  I suspect it's probably correct for most cases and might not be worth worrying about. AFAICT `npm repo` doesn't allow `@` as input (doesn't work with scope names) so this is why I avoided capturing it in the regex.

What do you think about making the execute silent? I find the 'press a key' prompt slightly annoying and other gx stuff doesn't seem to have it.

gabebw commented:
>AFAICT npm repo doesn't allow @ as input (doesn't work with scope names) so this is why I avoided capturing it in the regex.

Hmm, I just ran `npm repo @angular/cli` on the command line (version 6.9.0) and it opened https://github.com/angular/angular-cli. Does that fail for you?

>What do you think about making the execute silent?

Yes, good idea! Please do.

jdsutherland committed:
Make execute silent

jdsutherland commented:
~~D'oh, I was in an older env (npm 6.2.0) where it does fail. Using a newer version, I too see that scopes are supported.~~

I didn't look at this carefully enough as it does work in npm 6.2.0. I thought I had tried it a few different ways and that it wasn't working. For example doing something like:
❯ npm repo @angular
npm ERR! code EINVALIDTAGNAME
npm ERR! Invalid tag name "@angular": Tags may not have any characters that encodeURIComponent encodes.

I originally read this as meaning that @ is an invalid character. Anyway, scratch what I said about it not working.

jdsutherland committed:
Regex matches scoped along with nonscoped names

Examples:
Scoped: '@angular/core' => @angular/core
Nonscoped:'react-router-dom/Link' => react-router-dom
https://regex101.com/r/rmj9d1/3

jdsutherland commented:
The regex isn't a perfect validation of npm names but it might be good enough for our purposes here. There's few additional quirks that aren't mentioned at https://github.com/npm/validate-npm-package-name#valid-names such as no leading underscore - see https://docs.npmjs.com/files/package.json#name.

Do you think this is worth worrying about?

gabebw commented:
Re: invalid names, I think it's fine to not handle that case. If it's a big issue, I can look up how to catch exit statuses for commands in Vim and show an error `:echoerr`, but what's here is great for now.

Thanks for your help with this (and thanks for going through a few rounds of review on the PR) ✨. I'm going to squash and merge.
